File: foreign/go/internal/vsr/session.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package vsr

import (
	"crypto/rand"
	"encoding/binary"
	"errors"
	"math"
)

// Session errors. They are programming or protocol faults rather than server
// rejections, so they carry no IggyError code.
var (
	// ErrSessionAlreadyBound is returned when a second Bind arrives without an
	// intervening reset.
	ErrSessionAlreadyBound = errors.New("vsr: session already bound")
	// ErrInvalidSession is returned when a register reply carries a zero
	// session fence.
	ErrInvalidSession = errors.New("vsr: session must be greater than zero")
	// ErrRequestCounterExhausted is returned when the request watermark would
	// wrap past its u64 range.
	ErrRequestCounterExhausted = errors.New("vsr: request counter exhausted")
	// ErrNotBound is returned when a replicated request id is taken before
	// Register commits.
	ErrNotBound = errors.New("vsr: request id taken before the session is bound")
)

// Session is the consensus session state, a port of core/sdk/src/session.rs.
// It holds the ephemeral client identifier, the session fence the server
// assigns once Register commits, and the request watermark.
//
// Not safe for concurrent use. The transport owns one per connection and
// serializes access through its exchange lock, which the lockstep request
// model already requires.
type Session struct {
	client           ClientID
	session          uint64
	bound            bool
	requestCounter   uint64
	registerConsumed bool
}

// NewSession returns an unbound session with a fresh client identifier.
func NewSession() *Session {
	return NewSessionWithClientID(newClientID())
}

// NewSessionWithClientID returns an unbound session with a caller-supplied
// client identifier.
func NewSessionWithClientID(client ClientID) *Session {
	return &Session{client: client, requestCounter: 1}
}

// ClientID returns the current ephemeral client identifier.
func (s *Session) ClientID() ClientID {
	return s.client
}

// Bound reports whether Register has committed and the fence is set.
func (s *Session) Bound() bool {
	return s.bound
}

// SessionID returns the bound session fence, or 0 while unbound.
func (s *Session) SessionID() uint64 {
	return s.session
}

// HasActivity reports whether the session has registered, bound, or issued a
// replicated request.
func (s *Session) HasActivity() bool {
	return s.registerConsumed || s.bound || s.requestCounter > 1
}

// BeginRegister arms the session for a Register and returns its request id,
// always 0. A session that already registered or bound is replaced wholesale
// with a fresh client identifier, so a re-login encodes a clean Register
// instead of tripping the one-shot guard.
func (s *Session) BeginRegister() uint64 {
	if s.registerConsumed || s.bound {
		s.reset()
	}
	s.registerConsumed = true
	return 0
}

// Bind records the session fence the server assigned when Register committed.
func (s *Session) Bind(session uint64) error {
	if s.bound {
		return ErrSessionAlreadyBound
	}
	if session == 0 {
		return ErrInvalidSession
	}
	s.session = session
	s.bound = true
	return nil
}

// NextRequestID returns the next replicated-metadata request id and advances
// the watermark: 1, 2, 3, and so on.
func (s *Session) NextRequestID() (uint64, error) {
	if !s.bound {
		return 0, ErrNotBound
	}
	if s.requestCounter == math.MaxUint64 {
		return 0, ErrRequestCounterExhausted
	}
	id := s.requestCounter
	s.requestCounter++
	return id, nil
}

// CurrentRequestID returns the watermark without advancing it. Non-replicated
// and partition-plane requests use it because neither consults the client
// table: non-replicated requests route by transport identity, and the
// partition plane replicates in per-partition groups with no dedup table at
// all. The table accepts any id above the watermark with no contiguity
// requirement, so consuming one here would not gap anything; the invariant
// that matters is that every partition request on a session carries the id
// the next metadata operation will claim, and that a partition-plane replay
// is therefore at-least-once.
func (s *Session) CurrentRequestID() uint64 {
	return s.requestCounter
}

// Reset drops the session and mints a fresh client identifier. A disconnect
// invalidates the server-side fence, so the next Register must arrive under a
// new identity.
func (s *Session) Reset() {
	s.reset()
}

func (s *Session) reset() {
	s.client = newClientID()
	s.session = 0
	s.bound = false
	s.requestCounter = 1
	s.registerConsumed = false
}

// newClientID draws an ephemeral u128, retrying the one value the server
// rejects during header validation. crypto/rand.Read does not report an error
// as of Go 1.24. It panics if the system entropy source is unavailable.
func newClientID() ClientID {
	var buf [16]byte
	for {
		_, _ = rand.Read(buf[:])
		client := ClientID{
			Lo: binary.LittleEndian.Uint64(buf[0:8]),
			Hi: binary.LittleEndian.Uint64(buf[8:16]),
		}
		if !client.IsZero() {
			return client
		}
	}
}

File: foreign/node/src/wire/vsr/session.ts
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { randomBytes } from 'node:crypto';

const MAX_U64 = 0xFFFF_FFFF_FFFF_FFFFn;

/**
 * Consensus-level session state, ported from `core/sdk/src/session.rs`.
 *
 * Each client instance generates an ephemeral random `clientId` (u128).
 * After a Register commits, the server assigns a `session` number (commit op
 * number). Replicated metadata requests advance a monotonic request watermark.
 * Non-replicated and partition-plane requests reuse the current value because
 * the server only applies request sequencing to replicated metadata.
 */
export class ConsensusSession {
  private _clientId: bigint;
  private _session: bigint | null;
  private requestCounter: bigint;
  private registerConsumed: boolean;

  constructor(clientId?: bigint) {
    this._clientId = clientId ?? generateClientId();
    this._session = null;
    this.requestCounter = 1n;
    this.registerConsumed = false;
  }

  get clientId(): bigint {
    return this._clientId;
  }

  /** The bound session number, or null before Register commits. */
  get session(): bigint | null {
    return this._session;
  }

  get isBound(): boolean {
    return this._session !== null;
  }

  get hasActivity(): boolean {
    return this.registerConsumed ||
      this._session !== null ||
      this.requestCounter > 1n;
  }

  /** Binds the session after Register commits through consensus. */
  bind(session: bigint): void {
    if (this._session !== null)
      throw new Error(`session already bound (session=${this._session})`);
    if (session <= 0n)
      throw new Error('session must be > 0');
    this._session = session;
  }

  /**
   * Begins a registration, re-arming the session for a re-login. A prior
   * consumed or bound session is replaced wholesale (fresh client id,
   * unbound), so a repeat login encodes a clean Register instead of tripping
   * the one-shot guard. Returns the register request id, always 0.
   */
  beginRegister(): bigint {
    if (this.registerConsumed || this.isBound) {
      this._clientId = generateClientId();
      this._session = null;
      this.requestCounter = 1n;
      this.registerConsumed = false;
    }
    this.registerConsumed = true;
    return 0n;
  }

  /**
   * Next application request id, advancing the counter: 1, 2, 3, ...
   * Request 0 is reserved for Register.
   */
  nextRequestId(): bigint {
    if (!this.isBound)
      throw new Error('nextRequestId called before bind');
    if (this.requestCounter === MAX_U64)
      throw new RangeError('VSR request counter exhausted');
    const id = this.requestCounter;
    this.requestCounter += 1n;
    return id;
  }

  /** Current counter value without advancing it. */
  currentRequestId(): bigint {
    return this.requestCounter;
  }
}

/** Ephemeral random u128, non-zero (retry on the astronomically unlikely 0). */
const generateClientId = (): bigint => {
  for (;;) {
    const id = randomBytes(16).reduce(
      (acc, byte) => (acc << 8n) | BigInt(byte), 0n);
    if (id !== 0n) return id;
  }
};
